pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -149,206 +145,142 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	if errors.IsNotFound(err) {
 		// The resource no longer exists, in which case we stop processing.
 		c.Logger.Errorf("pipeline run %q in work queue no longer exists", key)
-		return nil
+		return controller.NewPermanentError(err)
 	} else if err != nil {
 		return err
 	}
 
 	// Don't modify the informer's copy.
 	pr := original.DeepCopy()
+
 	if !pr.HasStarted() {
+		// HasStarted is made false by InitializeConditions
 		pr.Status.InitializeConditions()
 		// In case node time was not synchronized, when controller has been scheduled to other nodes.
 		if pr.Status.StartTime.Sub(pr.CreationTimestamp.Time) < 0 {
 			c.Logger.Warnf("PipelineRun %s createTimestamp %s is after the pipelineRun started %s", pr.GetRunKey(), pr.CreationTimestamp, pr.Status.StartTime)
 			pr.Status.StartTime = &pr.CreationTimestamp
 		}
-		// start goroutine to track pipelinerun timeout only startTime is not set
+		// start goroutine to track pipelinerun timeout only if startTime is not set
 		go c.timeoutHandler.WaitPipelineRun(pr, pr.Status.StartTime)
-	} else {
-		pr.Status.InitializeConditions()
+		// Emit events. During the first reconcile the status of the PipelineRun may change twice
+		// from not Started to Started and then to Running, so we need to sent the event here
+		// and at the end of 'Reconcile' again.
+		// We also want to send the "Started" event as soon as possible for anyone who may be waiting
+		// on the event to perform user facing initialisations, such has reset a CI check status
+		afterCondition := pr.Status.GetCondition(apis.ConditionSucceeded)
+		events.Emit(c.Recorder, nil, afterCondition, pr)
 	}
 
-	// In case of reconcile errors, we store the error in a multierror, attempt
-	// to update, and return the original error combined with any update error
-	var merr *multierror.Error
+	if pr.IsDone() {
+		c.Logger.Infof("taskrun done : %s \n", pr.Name)
+		var merr *multierror.Error
 
-	switch {
-	case pr.IsDone():
 		// We may be reading a version of the object that was stored at an older version
 		// and may not have had all of the assumed default specified.
 		pr.SetDefaults(contexts.WithUpgradeViaDefaulting(ctx))
 
-		c.updatePipelineResults(ctx, pr)
-		if err := artifacts.CleanupArtifactStorage(pr, c.KubeClientSet, c.Logger); err != nil {
-			c.Logger.Errorf("Failed to delete PVC for PipelineRun %s: %v", pr.Name, err)
-			return err
+		// Completion time is set, so we don't need the timeout handler anymore
+		c.timeoutHandler.Release(pr)
+
+		// and sync the update to etcd
+		if err := c.updateTaskRunsStatusDirectly(pr); err != nil {
+			merr = multierror.Append(err, merr)
+			c.Logger.Errorf("Failed to update TaskRun status for PipelineRun %s: %v", pr.Name, err)
 		}
+		// Even if taskrun status updates failed, try to sync pipeline results at least
```

Comment:
i'm kind of confused by this comment, is this referring to the pipeline results feature or something else (it seems like something else since the function being called is about labels and annotations)? and i'm not sure why we would want to be trying to do the sync anyway, maybe a bit more info?